waiter/src/waiter/state/maintainer.clj

```diff
@@ -605,8 +608,9 @@
                                        (let [grace-period-secs (t/seconds (int (get service-description "grace-period-secs")))
                                              expiry-mins-int (int (get service-description "instance-expiry-mins"))
                                              expiry-mins (t/minutes expiry-mins-int)
-                                             instance-expired? (fn instance-expired? [instance]
-                                                                 (or (contains? (:flags instance) :expired)
+                                             instance-expired? (fn instance-expired? [{:keys [flags id] :as instance}]
+                                                                 (or (contains? flags :expired)
+                                                                     (ejection-expiry/instance-expired? ejection-expiry-tracker service-id id)
```

Comment:
Under normal circumstances the `:expired` flag's presence in the instance flags indicates the instance is expired and is observable via the scheduler's state endpoint.   We're expanding the use of `:expired` here and it may be surprising to see "expired" instances that aren't observable via inspection of the scheduler's, and I also wonder if there are other future (or present) consequences of the this conflation where we might actually prefer to distinguish between instances that are truly "expired" and instances that have breached a failed request threshold.  The scheduler is currently solely responsible for determination of an instance being "expired".  It seems a stretch to call an instance that has repeatedly failed requests something that has "expired".  I'd sooner call these instances "unhealthy" than I would "expired", but I realize you want to pick up the replacement logic already implemented for expired instances.  In summary, this feels like a method of implementation makes the tradeoff for expediency and potentially has long-term, hard-to-quantify ramifications.

Reply:
Agree, it is a trade-off.

Age-based instance expiry is also triggered via this logic and unknown to the scheduler.

Reply:
Also, I think initially the scheduled didn't know what was expired, we did only age-based computation. We expanded the meaning to allow the scheduler to provide feedback on what it thinks are expired instances.